rust/src/pool/mod.rs

```diff
@@ -0,0 +1,409 @@
+use std::{
+    ops::DerefMut,
+    sync::{Arc, LockResult, RwLock},
+};
+
+use crate::error::Error;
+
+/// An abstraction for a pool of thread-safe elements
+pub trait Pool {
+    /// The key type used to identify entries in the pool.
+    type Key;
+    /// The type of elements indexed by the pool.
+    type Type;
+
+    /// Retrieves an entry from the pool.
+    /// The concrete type returned by the get operation may not be the concrete type `Self::Type`
+    /// but instead a wrapper type which behaves like `Self::Type`. This abstraction allows for
+    /// the pool to store complex types which track the lifetime of stored objects, enabling
+    /// nested catching.
```

Comment:
```suggestion
    /// The concrete type returned by the get operation may not be the concrete type `Self::Type`
    /// but instead a wrapper type which behaves like `Self::Type`. This abstraction allows for
    /// the pool to store complex types which track the lifetime of stored objects, enabling
    /// nested caching.
```

Not sure what the last sentence is saying, but in general I feel like this description would better fit on the type not the method.

Reply:
The docstring refers to the `impl DerefMut<Target = Self::Type> ...`, not to the `PoolEntry`.
It would be nice to have an trait alias and move the comment to it. However, trait alias are experimental in stable rust